### ISSUE TITLE
Bigger batch size with less memory consumption

### ISSUE THREAD

**Issue by krxxxxxxxanc** [fundamentalvision/Deformable-DETR#5]
Hello,

I do fine tuning on a custom dataset and I want to try, if a bigger batch size results in a better performance. But my graphics card has not enough memory to do a batch size bigger than 1. So I looked up a solution, but the final performance drops and the network converges slower with higher batch sizes (the higher the batch size, the slower the convergence). 

My code is printed below. I know that the training time will be nearly the same with my code, but with performance I mean better Average Precision. I read that this code is exactly the same as real bigger batch sizes, except for batch normalization layers. Could the problem be the batch normalization layers in the ResNet backbone or would that have another impact?

I changed these lines:

https://github.com/fundamentalvision/Deformable-DETR/blob/dc33c7de943647fb306baed46105293dbefb589b/engine.py#L40-L68

To:

```python
optimizer.zero_grad()
# for samples, targets in metric_logger.log_every(data_loader, print_freq, header):
for i in metric_logger.log_every(range(len(data_loader)), print_freq, header):
    outputs = model(samples)
    loss_dict = criterion(outputs, targets)
    weight_dict = criterion.weight_dict
    losses = sum(loss_dict[k] * weight_dict[k] for k in loss_dict.keys() if k in weight_dict)

    # reduce losses over all GPUs for logging purposes
    loss_dict_reduced = utils.reduce_dict(loss_dict)
    loss_dict_reduced_unscaled = {f'{k}_unscaled': v
                                  for k, v in loss_dict_reduced.items()}
    loss_dict_reduced_scaled = {k: v * weight_dict[k]
                                for k, v in loss_dict_reduced.items() if k in weight_dict}
    losses_reduced_scaled = sum(loss_dict_reduced_scaled.values())

    loss_value = losses_reduced_scaled.item()

    if not math.isfinite(loss_value):
        print("Loss is {}, stopping training".format(loss_value))
        print(loss_dict_reduced)
        sys.exit(1)

    # optimizer.zero_grad()
    losses.backward()

    if (i+1)%mini_batchsize == 0:
        if max_norm > 0:
            grad_total_norm = torch.nn.utils.clip_grad_norm_(model.parameters(), max_norm)
        else:
            grad_total_norm = utils.get_total_grad_norm(model.parameters(), max_norm)
        optimizer.step()
        optimizer.zero_grad()
```

Thank you in advance!

**Comment by jackroos** [comment]
In fact, a bigger batch size not always produces a better result at the same epochs, especially when batch normalization is frozen, since bigger batch size means less iterations to update the parameters, which may harm the performance when the model doesn't sufficiently converge.
